iml-agent/src/daemon_plugins/postoffice.rs

```
@@ -0,0 +1,174 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    agent_error::ImlAgentError,
+    daemon_plugins::{DaemonPlugin, Output},
+    env,
+    http_comms::mailbox_client::send,
+};
+use futures::{
+    stream::{StreamExt as _, TryStreamExt},
+    Future, FutureExt,
+};
+use inotify::{Inotify, WatchDescriptor, WatchMask};
+use parking_lot::Mutex;
+use std::{
+    collections::{HashMap, HashSet},
+    pin::Pin,
+    sync::Arc,
+};
+use stream_cancel::{StreamExt, Trigger, Tripwire};
+use tokio::{fs, net::UnixListener};
+use tokio_util::codec::{BytesCodec, FramedRead};
+
+pub struct POWD(pub Option<WatchDescriptor>);
+
+pub struct PostOffice {
+    // individual mailbox socket listeners
+    routes: Arc<Mutex<HashMap<String, Trigger>>>,
+    inotify: Arc<Mutex<Inotify>>,
+    wd: Arc<Mutex<POWD>>,
+}
+
+pub fn create() -> impl DaemonPlugin {
+    PostOffice {
+        wd: Arc::new(Mutex::new(POWD(None))),
+        inotify: Arc::new(Mutex::new(
+            Inotify::init().expect("Failed to initialize inotify"),
+        )),
+        routes: Arc::new(Mutex::new(HashMap::new())),
+    }
+}
+
+/// Return socket address for a given mailbox
+pub fn socket_name(mailbox: &str) -> String {
+    let sock_dir = env::get_var("SOCK_DIR");
+    format!("{}/postman-{}.sock", sock_dir, mailbox)
+}
+
+impl std::fmt::Debug for PostOffice {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "PostOffice {{ {:?} }}", self.routes.lock().keys())
+    }
+}
+
+// Returned trigger should bed drop'd to cause route to stop
+fn start_route(mailbox: String) -> Trigger {
+    let (trigger, tripwire) = Tripwire::new();
+    let addr = socket_name(&mailbox);
+
+    let rc = async move {
+        // remove old unix socket
+        let _ = fs::remove_file(&addr).await;
```

This could fail silently